### PR TITLE
Form Field and Filter to parse ISO-8601 timestamps

### LINKED ISSUE #189: Add support for ISO 8601 format dates/datetimes
See [Filtering on Date Fields With `DjangoFilterBackend` Doesn't Quite Work. · Issue #1338 · tomchristie/django-rest-framework](https://github.com/tomchristie/django-rest-framework/issues/1338) over on DRF. 

It [looks like there's no fix coming in Django soon](https://code.djangoproject.com/ticket/23448) so we should add something here. 

:loudspeaker: @copitux — sorry for the interruption — Would we be able to incorporate your code from [Iso 8601 DateTimeFilter to Django Rest Framework and Django-Filter](https://gist.github.com/copitux/5773821)? (It doesn't provide licence info.)

### PR DESCRIPTION
**Pull request by carltongibson**
PR to Fix #189. 

Todo: 

- [x] Form field and Filter
- [x] Sufficient test cases 
- [x] Docs

### CODE REVIEW

**Comment by tomchristie** [comment]
@carltongibson feel free to give me a nudge in a few days if I've not helped out with review etc. :)

File: django_filters/fields.py [hunk 1]
```diff
@@ -71,3 +73,25 @@ def compress(self, data_list):
...
+    the ``datetime.strptime`` standard library
+
+    # ISO 8601: ``http://www.w3.org/TR/NOTE-datetime``
+
+    Based on Gist example by David Medina https://gist.github.com/copitux/5773821
+    """
+    ISO_8601 = 'iso-8601'
+    input_formats = [ISO_8601]
+
+    def strptime(self, value, format):
```
Comment (citizen-stig) [inline django_filters/fields.py]: I think it's better to rename ```format``` to ```dateime_format```, because it shadows built in function

**Comment by mcastle** [comment]
just tried this out and the new `IsoDateTimeFilter` works for me. Nice job!

**Comment by carltongibson** [comment]
@mcastle Thanks for the feedback. Everything _should_ work but, well, timezones... :-) 

I will add a few more test cases for this and write up the docs, and sort Django 1.4 — grrrrr... — this week. Time for a new release I think.

File: django_filters/fields.py (hunk 1 above)
Comment (carltongibson) [inline django_filters/fields.py]: This is just the method signature from the (Django) superclass. (But, yes, I see your point.)

**Comment by tomchristie** [comment]
Neat :smile: - will we be able to map to this by default from REST framework's filter class then?

**Comment by carltongibson** [comment]
@tomchristie — Yeah, why not. I'm just adding a test or two now and some docs (then I was going to ask you to review) 

I'll look at DRF and open a matching PR for that change (and if we time it right update to the new django-filter version)

**Comment by tomchristie** [comment]
Rockin'
Of course we'll need to make sure it falls back to the standard field if this one doesn't exist.

**Comment by carltongibson** [comment]
Having DRF use the new field would be a good integration test as well

**Comment by tomchristie** [comment]
Very true.

**Comment by carltongibson** [comment]
OK, @tomchristie — can you just cast your eye on this one. It's got some tests, some docs, and I'm inclined to stick. (Ta!)

I'll open a new ticket to introduce a DRF-friendly base FilterSet class: people want this, but they also want more JavaScript friendly query parameter parsing — and no doubt other things too :-) — so it'll be worth the extra indirection.

File: docs/ref/fields.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+Fields Reference
+================
+
+``IsoDateTimeField``
+~~~~~~~~~~~~~~~~~~~~
+
+Extends ``django.forms.DateTimeField`` to parse ISO 8601 formated dates.
```
Comment (tomchristie) [inline docs/ref/fields.txt]: Probably worth something like "additionally parse" - should make it clear that it's both ISO8601 *or* the existing formats.

**Comment by leonprou** [comment]
@carltongibson Can you give a complete example how to use `IsoDateTimeField`? I read the docs and couldn't understand. In my case I've got a django model with `DateTimeField` and a filter for it that extends `django_filters.FilterSet`.

**Comment by carltongibson** [comment]
@leonprou See the example in the docs: https://github.com/carltongibson/django-filter/blob/develop/docs/ref/filters.txt#L241-L247

If that's not enough you're going to have to explain what you're not able to get working.

**Comment by leonprou** [comment]
@carltongibson thank you for your response!

**Comment by leonprou** [comment]
@carltongibson thank you for the quick response!
the example did help, but now I struggle to use the `gte` tag, my filter looks like this:

```
class MyClass(django_filters.FilterSet):
    last_update_date = IsoDateTimeFilter()
    class Meta:
        model = MyClass
        fields = {
            'last_update_date': ['exact', 'gte']
        }

```
it works for exact date but not for `gte`

**Comment by carltongibson** [comment]
You can't use the metaclass and the explicit filter declaration. You need to explicitly define two filters, one for each lookup. Again, this is in the docs: https://django-filter.readthedocs.io/en/latest/usage.html#declaring-filters

**Comment by leonprou** [comment]
Thanks, it's working!

**Pull request merged**